ouroboros-consensus/src/Ouroboros/Consensus/Mempool/Impl.hs

```diff
@@ -96,49 +101,64 @@ forkSyncStateOnTipPointChange :: ( MonadAsync m
                               -> MempoolEnv m blk
                               -> m ()
 forkSyncStateOnTipPointChange registry menv = do
-  initialTipPoint <- atomically $ getTipPoint mpEnvChainDB
-  onEachChange registry id initialTipPoint (getTipPoint mpEnvChainDB) action
- where
-  action _ = atomically $ void $ implSyncState menv
-  MempoolEnv { mpEnvChainDB } = menv
+    initialTipPoint <- atomically $ getTipPoint mpEnvChainDB
+    onEachChange registry id initialTipPoint (getTipPoint mpEnvChainDB) action
+  where
+    action _tipPoint = implSyncState menv
+    MempoolEnv { mpEnvChainDB } = menv
 
 {-------------------------------------------------------------------------------
   Mempool Implementation
 -------------------------------------------------------------------------------}
 
--- TODO: This may return some transactions as invalid that aren't new. Remove?
 implAddTxs :: forall m blk. (MonadSTM m, ApplyTx blk)
            => MempoolEnv m blk
-           -> [(GenTxId blk, GenTx blk)]
-           -> m [(GenTx blk, ApplyTxErr blk)]
-implAddTxs mpEnv@MempoolEnv{mpEnvStateVar, mpEnvLedgerCfg} txs =
-    atomically $ do
-      ValidationResult {
-        vrBefore,
-        vrValid,
-        vrInvalid
-      } <- validateNew <$> validateIS mpEnv
-      writeTVar mpEnvStateVar IS { isTxs   = vrValid
-                                 , isTip   = vrBefore
-                                 }
-      return vrInvalid
+           -> [GenTx blk]
+           -> m [(GenTx blk, Maybe (ApplyTxErr blk))]
+implAddTxs mpEnv@MempoolEnv{mpEnvStateVar, mpEnvLedgerCfg, mpEnvTracer} txs = do
+    (removed, accepted, rejected, mempoolSize) <- atomically $ do
+      -- First sync the state, which might remove some transactions
+      syncRes@ValidationResult { vrInvalid = removed } <- validateIS mpEnv
+      -- Then validate the new transactions
+      let ValidationResult
+            { vrBefore, vrValid
+            , vrNewValid = accepted
+            , vrInvalid  = rejected
+            } = validateNew syncRes
+      writeTVar mpEnvStateVar IS { isTxs = vrValid, isTip = vrBefore }
+      mempoolSize <- getMempoolSize mpEnv
+      return (removed, accepted, rejected, mempoolSize)
+
+    trace $ TraceMempoolRemoveTxs   (map fst removed)  mempoolSize
+    trace $ TraceMempoolAddTxs      accepted           mempoolSize
+    trace $ TraceMempoolRejectedTxs (map fst rejected) mempoolSize
```

Comment:
I recall in a later patch we put `unless (null ...)` around these. So that's fine.